src/interfaces/protocols-configure.ts

```diff
@@ -79,24 +79,42 @@ export class ProtocolsConfigure extends Message<ProtocolsConfigureMessage> {
     // Traverse nested rule sets
     for (const rootRecordPath in definition.structure) {
       const rootRuleSet = definition.structure[rootRecordPath];
-      ProtocolsConfigure.validateRuleSet(rootRuleSet, rootRecordPath, globalRoles);
+
+      // gather $contextRoles
+      const contextRoles: string[] = [];
+      for (const childRecordType in rootRuleSet) {
+        if (childRecordType.startsWith('$')) {
+          continue;
+        }
+        const childRuleSet: ProtocolRuleSet = rootRuleSet[childRecordType];
+        if (childRuleSet.$contextRole) {
+          contextRoles.push(`${rootRecordPath}/${childRecordType}`);
+        }
+      }
+
+      ProtocolsConfigure.validateRuleSet(rootRuleSet, rootRecordPath, [...globalRoles, ...contextRoles]);
     }
   }
 
-  private static validateRuleSet(ruleSet: ProtocolRuleSet, protocolPath: string, globalRoles: string[]): void {
+  private static validateRuleSet(ruleSet: ProtocolRuleSet, protocolPath: string, roles: string[]): void {
```

Comment:
More for self-reminder, 'coz I always forget, the method feels every time I review it! LOL.

```suggestion
  /**
   * Validates the given rule set structure then recursively validates the structure all its nested child rule sets.
   */
  private static validateRuleSet(ruleSet: ProtocolRuleSet, protocolPath: string, roles: string[]): void {
```